### PR TITLE
Python 3.12 support

### PR DESCRIPTION
**Pull request by bluetech**
~~This is currently blocked on https://github.com/sissaschool/xmlschema/issues/342.~~

Blocked on python 3.12-beta.2 release in https://github.com/actions/python-versions/blob/main/versions-manifest.json

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -51,12 +52,13 @@ jobs:
           "ubuntu-py39",
           "ubuntu-py310",
           "ubuntu-py311",
+          "ubuntu-py312",
           "ubuntu-pypy3",
 
           "macos-py37",
-          "macos-py38",
```
Comment (nicoddemus) [inline .github/workflows/test.yml:57]: Probably good to leave a comment here explaining why we are skipping macos-py38.

**Comment by bluetech** [comment]
The xmlschema problem is fixed. Now there is a test failure on windows. I don't have windows but seems like a simple issue, will try to fix it.

```
______________ TestWINLocalPath.test_owner_group_not_implemented ______________

self = <test_local.TestWINLocalPath object at 0x0000026A748AE450>
path1 = local('C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pytest-of-unknown\\pytest-0\\path1')

    def test_owner_group_not_implemented(self, path1):
        with pytest.raises(NotImplementedError):
>           path1.stat().owner

testing\_py\test_local.py:1222: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_pytest._py.path.Stat object at 0x0000026A74A25C40>, name = 'owner'

    def __getattr__(self, name: str) -> Any:
>       return getattr(self._osstatresult, "st_" + name)
E       AttributeError: 'os.stat_result' object has no attribute 'st_owner'

.tox\py312\Lib\site-packages\_pytest\_py\path.py:216: AttributeError
=========================== short test summary info ===========================
FAILED testing/_py/test_local.py::TestWINLocalPath::test_owner_group_not_implemented - AttributeError: 'os.stat_result' object has no attribute 'st_owner'
```

File: .github/workflows/test.yml (hunk 1 above)
Comment (bluetech) [reply .github/workflows/test.yml:57]: I noticed macos-py311 was missing, so I added py312 and removed py38 to keep the job count the same. With the faster Python release cycle, it is starting to become a lot.

**Comment by bluetech** [comment]
The failure doesn't make sense to me.

Relevant code in `Stat`:

```py
class Stat:
    ...

    def __getattr__(self, name: str) -> Any:
        return getattr(self._osstatresult, "st_" + name)

    def __init__(self, path, osstatresult):
        self.path = path
        self._osstatresult = osstatresult

    @property
    def owner(self):
        if iswin32:
            raise NotImplementedError("XXX win32")
        import pwd

        entry = error.checked_call(pwd.getpwuid, self.uid)  # type:ignore[attr-defined]
        return entry[0]
    ...
```

The failing test (`py.path.local` tests):

```py
    def test_owner_group_not_implemented(self, path1):
        with pytest.raises(NotImplementedError):
            path1.stat().owner
```

fails as seen in previous comment.

How come the `__getattr__` is reached for `owner` attribute, when there is an `owner` property? It only happens on Windows so I'm assuming it has to do with the `raise NotImplementedError("XXX win32")` code in the property (but the property isn't present in the stack trace...).

I reran the job and it happened again so it's reproducible.

**Comment by nicoddemus** [comment]
You are right, it does not make much sense. I debugged it, and here is what happens in 3.12:

1. It reaches `owner`, and raises `NotImplementedError` as expected;
2. At this points, somehow `__getattr__` gets called.

In Python 3.10 (and I supposed 3.11 but I did not test), 2) never happens, `NotImplementedError` is raised from the property, and `__getattr__` does not get called.

Seems like a bug in CPython, I will post an issue upstream.

**Comment by nicoddemus** [comment]
This has been fixed already and will be available on the next release. 👍

**Comment by bluetech** [comment]
Hmm some autocommit hook deletes the classifier so I'll drop it for now.

**Comment by RonnyPfannschmidt** [comment]
@bluetech you can probably up the max version in the setup.cfg firmat hook config

**Comment by hugovk** [comment]
Yes, it's this: https://github.com/asottile/setup-cfg-fmt#adds-python-version-classifiers

```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index af6cd262..fbd5d39c 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -50,7 +50,7 @@ repos:
     rev: v2.2.0
     hooks:
     -   id: setup-cfg-fmt
-        args: ["--max-py-version=3.11", "--include-version-classifiers"]
+        args: ["--max-py-version=3.12", "--include-version-classifiers"]
 -   repo: https://github.com/pre-commit/pygrep-hooks
     rev: v1.10.0
     hooks:
```

**Comment by bluetech** [comment]
Thanks, missed that.

**Comment by bluetech** [comment]
github updated to Python 3.12 beta, I reran the job and now there is a new deprecation (showing one, rest are the same):

```
 _________________ ERROR collecting testing/python/metafunc.py __________________
testing/python/metafunc.py:16: in <module>
    import hypothesis
<frozen importlib._bootstrap>:1266: in _find_and_load
    ???
<frozen importlib._bootstrap>:1237: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:841: in _load_unlocked
    ???
.tox/py312-coverage/lib/python3.12/site-packages/_pytest/assertion/rewrite.py:169: in exec_module
    source_stat, co = _rewrite_test(fn, self.config)
.tox/py312-coverage/lib/python3.12/site-packages/_pytest/assertion/rewrite.py:352: in _rewrite_test
    rewrite_asserts(tree, source, strfn, config)
.tox/py312-coverage/lib/python3.12/site-packages/_pytest/assertion/rewrite.py:413: in rewrite_asserts
    AssertionRewriter(module_path, config, source).run(mod)
.tox/py312-coverage/lib/python3.12/site-packages/_pytest/assertion/rewrite.py:691: in run
    doc = item.value.s
/opt/hostedtoolcache/Python/3.12.0-beta.1/x64/lib/python3.12/ast.py:532: in _s_getter
    warnings._deprecated(
/opt/hostedtoolcache/Python/3.12.0-beta.1/x64/lib/python3.12/warnings.py:529: in _deprecated
    warn(msg, DeprecationWarning, stacklevel=3)
E   DeprecationWarning: Attribute s is deprecated and will be removed in Python 3.14; use value instead
=========================== short test summary info ============================
ERROR testing/test_error_diffs.py - DeprecationWarning: Attribute s is deprecated and will be removed in Python...
ERROR testing/test_meta.py - DeprecationWarning: Attribute s is deprecated and will be removed in Python...
ERROR testing/test_runner_xunit.py - DeprecationWarning: Attribute s is deprecated and will be removed in Python...
ERROR testing/test_terminal.py - DeprecationWarning: Attribute s is deprecated and will be removed in Python...
ERROR testing/python/metafunc.py - DeprecationWarning: Attribute s is deprecated and will be removed in Python...
```

Fixing.

**Comment by bluetech** [comment]
Pushed a fix for `imp` module removal. Remaining two failures are in assertion rewriting:

```

**Comment by bluetech** [comment]
cc @asottile on previous comment, you know tokenization better than anyone, maybe you have an idea :)

**Comment by asottile** [comment]
yeah that's all to be expected. the DEDENT might change back to be less disruptive in https://github.com/python/cpython/issues/104976

**Comment by bluetech** [comment]
Thanks, I see the PR is merged now, so let's wait for the next beta.

**Comment by hugovk** [comment]
> Thanks, I see the PR is merged now, so let's wait for the next beta.

3.12.0b2 is now out:

* https://www.python.org/downloads/release/python-3120b2/

**Comment by bluetech** [comment]
~~Even though 3.12-beta.2 appears in https://github.com/actions/python-versions/releases, I can't get it to be picked up by the `setup-python` action. I tried setting `check-latest: true` but this doesn't work either. I'm assuming it needs more time. If anyone knows how to speed it up, let me know.~~

I see now that it needs to appear in https://github.com/actions/python-versions/blob/main/versions-manifest.json not in the releases.

**Comment by The-Compiler** [comment]
Yup, I believe https://github.com/actions/python-versions/pull/232 is the thing to watch.

**Comment by hugovk** [comment]
https://github.com/actions/python-versions/pull/232 has been merged and beta 2 is now available ([for example](https://github.com/hugovk/norwegianblue/actions/runs/5140630939/jobs/9425773257#step:3:39)).

**Comment by bluetech** [comment]
OK this is good to go now, marked as ready for review.

**Comment by bluetech** [comment]
(I changed the branch protection checks already so this can be green)

**Review by Pierre-Sassoulas — APPROVED** [review]
Great work ! Just curious, I wonder if the python 3.7 specific code path will be removed in 2 weeks for python 3.7 eol ? Is there also a 'cleanup' of the code when removing support for an interpreter ?

**Comment by bluetech** [comment]
> Great work ! Just curious, I wonder if the python 3.7 specific code path will be removed in 2 weeks for python 3.7 eol ? Is there also a 'cleanup' of the code when removing support for an interpreter ?

Yes, I believe our policy is that new pytest feature releases only support non-EOL python versions. Then we also clean up all of the compat code for that version.

**Pull request merged**